src/graph/network.cc

```diff
@@ -194,182 +267,312 @@ DGL_REGISTER_GLOBAL("network._CAPI_SenderSendNodeFlow")
+        {meta.data_shape_[9]},
+        DLDataType{kDLInt, 64, 1},
+        DLContext{kDLCPU, 0},
+        array_4.data);
       // CSR indice
       Message array_5;
-      CHECK_NE(receiver->RecvFrom(&array_5, send_id), -1);
-      CHECK_EQ(msg.data_shape_[10], 1);
-      DLTensor indice_tensor;
-      ConstructNFTensor(&indice_tensor, array_5.data, msg.data_shape_[11]);
-      DLManagedTensor *indice_managed_tensor = new DLManagedTensor();
-      indice_managed_tensor->dl_tensor = indice_tensor;
-      NDArray indice = NDArray::FromDLPack(indice_managed_tensor);
+      CHECK_EQ(receiver->RecvFrom(&array_5, send_id), REMOVE_SUCCESS);
+      CHECK_EQ(meta.data_shape_[10], 1);
+      NDArray indice = CreateNDArrayFromRaw(
+        {meta.data_shape_[11]},
+        DLDataType{kDLInt, 64, 1},
+        DLContext{kDLCPU, 0},
+        array_5.data);
       // CSR edge_ids
       Message array_6;
-      CHECK_NE(receiver->RecvFrom(&array_6, send_id), -1);
-      CHECK_EQ(msg.data_shape_[12], 1);
-      DLTensor edge_id_tensor;
-      ConstructNFTensor(&edge_id_tensor, array_6.data, msg.data_shape_[13]);
-      DLManagedTensor *edge_id_managed_tensor = new DLManagedTensor();
-      edge_id_managed_tensor->dl_tensor = edge_id_tensor;
-      NDArray edge_ids = NDArray::FromDLPack(edge_id_managed_tensor);
+      CHECK_EQ(receiver->RecvFrom(&array_6, send_id), REMOVE_SUCCESS);
+      CHECK_EQ(meta.data_shape_[12], 1);
+      NDArray edge_ids = CreateNDArrayFromRaw(
+        {meta.data_shape_[13]},
+        DLDataType{kDLInt, 64, 1},
+        DLContext{kDLCPU, 0},
+        array_6.data);
+
       // Create CSR
       CSRPtr csr(new CSR(indptr, indice, edge_ids));
       nf->graph = GraphPtr(new ImmutableGraph(csr, nullptr));
       *rv = nf;
-    } else if (msg.msg_type() == kEndMsg) {
-      *rv = msg.msg_type();
+    } else if (meta.msg_type() == kEndMsg) {
+      *rv = meta.msg_type();
     } else {
-      LOG(FATAL) << "Unknown message type: " << msg.msg_type();
+      LOG(FATAL) << "Unknown message type: " << meta.msg_type();
     }
   });
 
+
+////////////////////////// Distributed KVStore Components ////////////////////////////////
+
+
+DGL_REGISTER_GLOBAL("network._CAPI_SenderSendKVMsg")
+.set_body([] (DGLArgs args, DGLRetValue* rv) {
+    CommunicatorHandle chandle = args[0];
+    int recv_id = args[1];
+    KVStoreMsg kv_msg;
+    kv_msg.msg_type = args[2];
+    kv_msg.rank = args[3];
+    network::Sender* sender = static_cast<network::Sender*>(chandle);
+    if (kv_msg.msg_type != kEndMsg && kv_msg.msg_type != kBarrierMsg) {
+      std::string name = args[4];
+      kv_msg.name = name;
+      kv_msg.id = args[5];
+      if (kv_msg.msg_type != kPullMsg) {
+        kv_msg.data = args[6];
+      }
+    }
+    int64_t kv_size = 0;
+    char* kv_data = kv_msg.Serialize(&kv_size);
+    // Send kv_data
+    Message send_kv_msg;
+    send_kv_msg.data = kv_data;
+    send_kv_msg.size = kv_size;
+    send_kv_msg.deallocator = DefaultMessageDeleter;
```

Comment:
I don't understand. Here it seems you send data. What is the data message below. Please document every message here. Why are there so many messages?

Reply:
We have two types of messages here. `KVStoreMsg` is a wrapper for python message. `Message` is used for communicator, which is a pure binary message.